gateway/gateway-requests/src/registration/handshake/state.rs

```diff
@@ -15,9 +16,57 @@ use futures::{Sink, SinkExt, Stream, StreamExt};
 use log::*;
 use nymsphinx::params::{GatewayEncryptionAlgorithm, GatewaySharedKeyHkdfAlgorithm};
 use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
 use std::convert::{TryFrom, TryInto};
 use tungstenite::Message as WsMessage;
 
+#[derive(Serialize, Deserialize)]
+pub struct InitMessage {
+    local_id_pubkey: Vec<u8>,
+    ephemeral_key: Vec<u8>,
+    credential: Option<Credential>,
+}
+
+impl InitMessage {
+    fn new(
+        local_id_pubkey: &identity::PublicKey,
+        ephemeral_key: &encryption::PublicKey,
+        credential: Credential,
+    ) -> Self {
+        InitMessage {
+            local_id_pubkey: local_id_pubkey.to_bytes().to_vec(),
+            ephemeral_key: ephemeral_key.to_bytes().to_vec(),
+            credential: Some(credential),
+        }
+    }
+
+    pub fn local_id_pubkey(&self) -> identity::PublicKey {
+        identity::PublicKey::from_bytes(&self.local_id_pubkey).unwrap()
+    }
+
+    pub fn ephemeral_key(&self) -> encryption::PublicKey {
+        encryption::PublicKey::from_bytes(&self.ephemeral_key).unwrap()
+    }
+
+    pub fn credential(&self) -> Credential {
+        self.credential.clone().unwrap()
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        bincode::serialize(self).unwrap()
```

Comment:
I'm not entirely sure what to think about the use of bincode here. Sure, it is way cleaner this way, but is the sucker 100% deterministic all the time? Are we not signing the result of the sucker?

Reply:
I would say `bincode` _should_ be deterministic, at least for the struct that we're using now (with `Vec<u8>` and `Option<Credential>`). The only way I can see it potentially become indeterministic like `protobuf` is to have some map-like structure in there. But even then, I don't fully see the problem, as long as the deserialization uses the appropriate algorithm and knows how to get the struct out of the bytes, the message exchange should be successful, right?